### PR TITLE
Compliant dex

### PR DESCRIPTION
**Pull request by anhdungle93**
Add a ZkKYC compliant version of UniswapV2:
1) add a registerKYC function to the router contract so user can mint SBT.
2) modify router functions to require SBT.
3) restrict UniswapPair functions to only be callable by the router so the SBT requirement is enforced.

### CODE REVIEW

**Review by zwilling — CHANGES_REQUESTED** [review]
I have a couple conceptual questions to the compliant DEX. They might also be interesting for @mike-tis . As long as this is only an example it is probably ok. They just need to be resolved before someone creates a real DEX out of it.

1. If we want to provide a better DEX, we can already upgrade to Uniswap V3. It is under GPL since April 2023. Is this something we want to consider?
2. Because users of the DEX are required to have passed the requirements it is harder to integrate the DEX in other DeFi smart contracts. It would be good to have a concept or example how a smart contract can be whitelisted to interact with the DEX.
3. Adding and removing liquidity requires the SBT, but transferring the liquidity token does not. Is this compliant enough?
4. In general, someone could set up a proxy layer on top of the DEX that. That someone needs to pass the zkKYC to the proxy. Afterwards non-compliant users could access the DEX through the proxy. The Router does support zkKYC fraud investigation, but there is nothing to counter the non-compliant activity, such as a blacklist.
5. The operator of the DEX can change the router contract and with it the compliance requirements. Thus the operator can also freeze user funds in the DEX. Is that what we want? The problem might be that such a control might be considered as custody meaning that the DEX might become a financial intermediary that needs to register and be regulated. This might be ok because we are building DEX features to actually comply with regulation. But is is ok to launch such a DEX on mainnet? I do not have the expertise for those legal questions, but [GPT4o shares my concern](https://chatgpt.com/share/67176cf1-0ce8-800b-95ab-3b1f438f8fcd). We could maybe allow users that have once passed the compliance to remove liquidity without going through the router again. This might remove the concern about the custody to freeze user assets.

Feedback on the PR itself:
1. The code looks good.
2. I am not sure if the Compliant DEX code fits into the zkKYC package. If it is meant as example, ok. But if it should become a real product it would be better to put it into another package. Otherwise the zkCertificates package would get blown up with dApps of various complexity.
3. Would be good to have some deployment scripts for testnet.
4. I am missing some kind of license acknolwledgement for the DEX. Something like a readme in `contracts/dapps/CompliantDex` stating the source and license.

**Comment by mike-tis** [comment]
> I have a couple conceptual questions to the compliant DEX. They might also be interesting for @mike-tis . As long as this is only an example it is probably ok. They just need to be resolved before someone creates a real DEX out of it.

Thanks Frederik! And yes, it's a demo.

> 1. If we want to provide a better DEX, we can already upgrade to Uniswap V3. It is under GPL since April 2023. Is this something we want to consider?

Because it's a demo, it is simpler to use v2.

> 2. Because users of the DEX are required to have passed the requirements it is harder to integrate the DEX in other DeFi smart contracts. It would be good to have a concept or example how a smart contract can be whitelisted to interact with the DEX.

Good suggestion, but I think we can keep it for later.

> 3. Adding and removing liquidity requires the SBT, but transferring the liquidity token does not. Is this compliant enough?

I would vote for making LTs compliant using the code you wrote for the previous demo. But can be done as a feature in some time I think.

> 4. In general, someone could set up a proxy layer on top of the DEX that. That someone needs to pass the zkKYC to the proxy. Afterwards non-compliant users could access the DEX through the proxy. The Router does support zkKYC fraud investigation, but there is nothing to counter the non-compliant activity, such as a blacklist.

It is interesting, do you have any ideas how we can prevent this?

> 5. The operator of the DEX can change the router contract and with it the compliance requirements. Thus the operator can also freeze user funds in the DEX. Is that what we want? The problem might be that such a control might be considered as custody meaning that the DEX might become a financial intermediary that needs to register and be regulated. This might be ok because we are building DEX features to actually comply with regulation. But is is ok to launch such a DEX on mainnet? I do not have the expertise for those legal questions, but [GPT4o shares my concern](https://chatgpt.com/share/67176cf1-0ce8-800b-95ab-3b1f438f8fcd). We could maybe allow users that have once passed the compliance to remove liquidity without going through the router again. This might remove the concern about the custody to freeze user assets.

For prod, I would vote to make it non-upgradable, but now we are not even planning to push it to the prod.
To allow withdrawals without a verification - good thing, I would say we should do it.

> 2. I am not sure if the Compliant DEX code fits into the zkKYC package. If it is meant as example, ok. But if it should become a real product it would be better to put it into another package. Otherwise the zkCertificates package would get blown up with dApps of various complexity.

Better to make a separate repo I think.

**Comment by zwilling** [comment]
>> In general, someone could set up a proxy layer on top of the DEX that. That someone needs to pass the zkKYC to the proxy. Afterwards non-compliant users could access the DEX through the proxy. The Router does support zkKYC fraud investigation, but there is nothing to counter the non-compliant activity, such as a blacklist.
> It is interesting, do you have any ideas how we can prevent this?
We could additionally check that the `tx.origin` has a SBT. Then the submitter of the tx is also compliant. But it has some limitations because compliant users can be tricked into calling solidity code, that executes non-compliant stuff in the middle. https://docs.soliditylang.org/en/v0.8.28/security-considerations.html#tx-origin

File: packages/zk-certificates/test/contracts/CompliantUniswapV2.ts [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+    const { proof, publicSignals } = await groth16.fullProve(
+      input,
+      wasmPath,
+      zkeyPath,
+    );
+
+    const [piA, piB, piC] = processProof(proof);
+    const publicInputs = processPublicSignals(publicSignals);
+
+    const publicTime = parseInt(publicSignals[6], 10); // Assuming INDEX_CURRENT_TIME is 1
```
[the remark this answers is not in the document]
Comment (anhdungle93) [reply packages/zk-certificates/test/contracts/CompliantUniswapV2.ts]: comment fixed

File: packages/zk-certificates/contracts/dapps/CompliantDex/WETH9.sol [hunk 2]
```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+pragma solidity ^0.8.0;
+
+contract WETH9 {
```
[the remark this answers is not in the document]
Comment (anhdungle93) [reply packages/zk-certificates/contracts/dapps/CompliantDex/WETH9.sol:5]: that would be good, can you publish the repo?

[the remark this answers is not in the document]
Comment (anhdungle93) [reply packages/zk-certificates/contracts/dapps/CompliantDex/WETH9.sol:5]: that would be good, can you publish the repo?

File: packages/zk-certificates/contracts/dapps/CompliantDex/UniswapV2Router02.sol [hunk 3]
```diff
@@ -0,0 +1,523 @@
+pragma solidity ^0.8.0;
```
[the remark this answers is not in the document]
Comment (anhdungle93) [reply packages/zk-certificates/contracts/dapps/CompliantDex/UniswapV2Router02.sol:3]: I added the license

File: packages/zk-certificates/hardhat.config.ts [hunk 4]
```diff
@@ -25,14 +25,39 @@ const config: HardhatUserConfig = {
...
+            runs: 100,
+          },
+          viaIR: true,
+        },
+      },
+      {
+        version: '0.5.6',
       },
     ],
+    viaIR: true,
```
[the remark this answers is not in the document]
Comment (anhdungle93) [reply packages/zk-certificates/hardhat.config.ts:60]: yeah, it seems like I tried but could not set it globally.

**Comment by anhdungle93** [comment]
> I would vote for making LTs compliant using the code you wrote for the previous demo. But can be done as a feature in some time I think.
> 
I can modify the LT to follow the CompliantERC20

> For prod, I would vote to make it non-upgradable, but now we are not even planning to push it to the prod. To allow withdrawals without a verification - good thing, I would say we should do it.
> 
I will remove verification regarding withdrawals

> 
> Better to make a separate repo I think.
I will make a new repo

**Comment by anhdungle93** [comment]
> > > In general, someone could set up a proxy layer on top of the DEX that. That someone needs to pass the zkKYC to the proxy. Afterwards non-compliant users could access the DEX through the proxy. The Router does support zkKYC fraud investigation, but there is nothing to counter the non-compliant activity, such as a blacklist.
> 
> > It is interesting, do you have any ideas how we can prevent this?
> 
> We could additionally check that the `tx.origin` has a SBT. Then the submitter of the tx is also compliant. But it has some limitations because compliant users can be tricked into calling solidity code, that executes non-compliant stuff in the middle. https://docs.soliditylang.org/en/v0.8.28/security-considerations.html#tx-origin

I think the limitation is not that bad, because
1) User needs to use a contract wallet in the first place
2) the usage of tx.origin is malicious, which we can prove in our casethat it's not by having our code verified and advanced users using that "flawed" contract wallet should always check the smart contract that they interact with.
On the other hand using `tx.origin` still doesn't fully eliminate the problem, because the SBT can reside in an externally owned account and a proxy can be deployed ony taking tx from this EOA. The exploiter reveals the private key of this EOA to anyone, then when any user needs to make a trade, he approves the proxy and makes a signed message. User submits the signs message from EOA to the proxy, the proxy checks the signature, takes fund from user's address, makes the trade then returns the output to user's address.

Another stricter solution would be requiring that only EOA can interact with the router contract, which this limits composability.

File: packages/zk-certificates/contracts/dapps/CompliantDex/WETH9.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (zwilling) [reply packages/zk-certificates/contracts/dapps/CompliantDex/WETH9.sol:5]: Sure, here it is: https://www.npmjs.com/package/@galactica-net/wrapped-gnet?activeTab=readme

**Comment by zwilling** [comment]
Thanks a lot @anhdungle93 .
I close this PR because the Compliant DEX is now available here: https://github.com/Galactica-corp/Compliant-DEX

**Pull request closed without merging**